plugins/Live/Model.php

```diff
@@ -43,20 +43,123 @@ public function queryLogVisits($idSite, $period, $date, $segment, $offset, $limi
+        // If no other filter, only look at the last 24 hours of stats
+        if (empty($visitorId)
+            && empty($limit)
+            && empty($offset)
+            && empty($period)
+            && empty($date)
+        ) {
+            $period = 'day';
+            $date = 'yesterdaySameTime';
+        }
+
+        list($dateStart, $dateEnd) = $this->getStartAndEndDate($idSite, $period, $date);
+
+        $virtualDateEnd = $dateEnd;
+        if (empty($dateEnd)) {
+            $virtualDateEnd = Date::now()->addDay(1); // matomo always adds one day for some reason
+        }
+
+        $virtualDateStart = $dateStart;
+        if (empty($virtualDateStart)) {
+            $virtualDateStart = Date::factory(Date::FIRST_WEBSITE_TIMESTAMP); // matomo always adds one day for some reason
+        }
+
+        $queries = [];
+        $hasStartEndDateMoreThanOneDayInBetween = $virtualDateStart && $virtualDateStart->addDay(1)->isEarlier($virtualDateEnd);
+        if ($limit
+            && $hasStartEndDateMoreThanOneDayInBetween
+        ) {
+            if ($hasStartEndDateMoreThanOneDayInBetween) {
+                $virtualDateEnd = $virtualDateEnd->subDay(1);
+                $queries[] = array($virtualDateEnd, $dateEnd); // need to use ",endDate" in case endDate is not set
+            }
+
+                if ($virtualDateStart->addDay(7)->isEarlier($virtualDateEnd)) {
+                    $queries[] = array($virtualDateEnd->subDay(7), $virtualDateEnd->subSeconds(1));
+                    $virtualDateEnd = $virtualDateEnd->subDay(7);
+                }
+
+            if (!$offset) {
+                // only when no offset
+                // we would in worst case - if not enough visits are found to bypass the offset - execute below queries too often.
+                // like we would need to execute each of the queries twice just to find out if there are some visits that
+                // need to be skipped...
+
+                if ($virtualDateStart->addDay(30)->isEarlier($virtualDateEnd)) {
+                    $queries[] = array($virtualDateEnd->subDay(30), $virtualDateEnd->subSeconds(1));
+                    $virtualDateEnd = $virtualDateEnd->subDay(30);
+                }
+                if ($virtualDateStart->addPeriod(1, 'year')->isEarlier($virtualDateEnd)) {
+                    $queries[] = array($virtualDateEnd->subYear(1), $virtualDateEnd->subSeconds(1));
+                    $virtualDateEnd = $virtualDateEnd->subYear(1);
+                }
+            }
+
+            if ($virtualDateStart->isEarlier($virtualDateEnd)) {
+                // need to use ",endDate" in case startDate is not set in which case we do not want to have any limit
+                $queries[] = array($dateStart, $virtualDateEnd->subSeconds(1));
+            }
+        } else {
+            $queries[] = array($dateStart, $dateEnd);
+        }
```

Comment:
Here's an idea that might make it a bit faster (assuming I'm reading this correctly). What the code is doing now is this:

- check dateStart => dateStart + 1 day
- still no visits? check dateStart => dateStart + 7 days
- still no visits? check dateStart => dateStart + 30 days
- ... etc.

In each query we're checking the same rows as the previous one, we could change it to something like:

- check dateStart => dateStart + 1 day
- still no visits? check dateStart + 1 day => dateStart + 7 days
- still no visits? check dateStart + 7 days => dateStart + 30 days
- ... etc.

Reply:
@diosmosis it should already do what you suggested. Unless I'm not understanding it right. But basically the code does not fetch the same day or week twice for example.

Reply:
Ah ok, `$virtualEndDate` changes, I didn't notice that.